File: 백준/Gold/2580. 스도쿠/스도쿠.py
import sys

input = sys.stdin.readline
write = sys.stdout.write

def solve(cnt: int, zero: list[tuple], nlist: list[list[int]], row_check: list[list[bool]], col_check: list[list[bool]], box_check: list[list[bool]]):
    if cnt == len(zero):
        for i in nlist:
            print(*i)
        sys.exit(0)
    
    r, c = zero[cnt]
    box_idx = (r // 3) * 3 + (c // 3)
    
    for num in range(1, 10):
        if not row_check[r][num] and not col_check[c][num] and not box_check[box_idx][num]:
            row_check[r][num] = col_check[c][num] = box_check[box_idx][num] = True
            nlist[r][c] = num
            
            solve(cnt + 1, zero, nlist, row_check, col_check, box_check)
            
            nlist[r][c] = 0
            row_check[r][num] = col_check[c][num] = box_check[box_idx][num] = False
        
def main():
    nlist = []
    row_check = [[False] * 10 for _ in range(9)]
    col_check = [[False] * 10 for _ in range(9)]
    box_check = [[False] * 10 for _ in range(9)]
    zero = []
    
    for i in range(9):
        row = list(map(int, input().split()))
        nlist.append(row)
        for j in range(9):
            if row[j] != 0:
                num = row[j]
                row_check[i][num] = True
                col_check[j][num] = True
                box_check[((i // 3) * 3) + (j // 3)][num] = True
            else:
                zero.append((i, j))
                
    solve(0, zero, nlist, row_check, col_check, box_check)
                
if __name__ == "__main__":
    main()

File: 백준/Gold/2580. 스도쿠/스도쿠.cc
#include <stdio.h>
#include <stdlib.h>

int board[9][9];
int row_check[9][10];
int col_check[9][10]; 
int box_check[9][10];

typedef struct {
    int r, c;
} Point;

Point zero[81];
int zero_cnt = 0;

int solve(int cnt) {
    if (cnt == zero_cnt) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                printf("%d ", board[i][j]);
            }
            printf("\n");
        }
        return 1;
    }

    int r = zero[cnt].r;
    int c = zero[cnt].c;
    int box_idx = (r / 3) * 3 + (c / 3);

    for (int num = 1; num <= 9; num++) {
        if (!row_check[r][num] && !col_check[c][num] && !box_check[box_idx][num]) {
            row_check[r][num] = col_check[c][num] = box_check[box_idx][num] = 1;
            board[r][c] = num;

            if (solve(cnt + 1)) return 1;

            board[r][c] = 0;
            row_check[r][num] = col_check[c][num] = box_check[box_idx][num] = 0;
        }
    }
    return 0;
}

int main(void) {
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
            if (scanf("%d", &board[i][j]) != 1) return 0;

            if (board[i][j] != 0) {
                int num = board[i][j];
                row_check[i][num] = 1;
                col_check[j][num] = 1;
                box_check[(i / 3) * 3 + (j / 3)][num] = 1;
            }
            else {
                zero[zero_cnt].r = i;
                zero[zero_cnt].c = j;
                zero_cnt++;
            }
        }
    }

    solve(0);

    return 0;
}